torch/csrc/inductor/static_cuda_launcher.cpp

```diff
@@ -0,0 +1,502 @@
+static const at::cuda::NVRTC& nvrtc() {
+  return at::globalContext().getNVRTC();
+}
+
+static CUdeviceptr getPointer(PyObject* obj) {
+  CUdeviceptr data_ptr = 0;
+  if (THPUtils_checkLong(obj)) {
+    data_ptr = THPUtils_unpackUInt64(obj);
+    return data_ptr;
+  }
+  if (obj == Py_None) {
+    // valid nullptr
+    return data_ptr;
+  }
+  PyObject* ptr = PyObject_GetAttrString(obj, "data_ptr");
+  if (ptr) {
+    PyObject* empty_tuple = PyTuple_New(0);
+    PyObject* ret = PyObject_Call(ptr, empty_tuple, nullptr);
+    Py_DECREF(empty_tuple);
+    Py_DECREF(ptr);
+    if (!THPUtils_checkLong(ret)) {
+      throw std::runtime_error(
+          "data_ptr method of Pointer object must return 64-bit int");
+    }
+    data_ptr = THPUtils_unpackUInt64(ret);
+    if (!data_ptr)
+      return data_ptr;
+
+    CUdeviceptr dev_ptr = 0;
+    AT_CUDA_DRIVER_CHECK(nvrtc().cuPointerGetAttribute(
+        &dev_ptr, CU_POINTER_ATTRIBUTE_DEVICE_POINTER, data_ptr));
+    Py_DECREF(ret);
+    return dev_ptr;
+  }
+  throw std::runtime_error(
+      "Pointer argument must be either uint64 or have data_ptr method");
+}
+
+static inline CUfunction loadKernel(
+    std::string filePath,
+    const std::string& funcName,
+    uint32_t sharedMemBytes,
+    const std::optional<std::string>& cubinDir = std::nullopt) {
+  if (cubinDir) {
+    std::filesystem::path p1{*cubinDir};
+    std::filesystem::path p2{filePath};
+    filePath = (p1 / p2.filename()).string();
+  }
+
+  CUmodule mod = nullptr;
+  CUfunction func = nullptr;
+  AT_CUDA_DRIVER_CHECK(nvrtc().cuModuleLoad(&mod, filePath.c_str()));
+  AT_CUDA_DRIVER_CHECK(
+      nvrtc().cuModuleGetFunction(&func, mod, funcName.c_str()));
+  if (sharedMemBytes > 0) {
+    AT_CUDA_DRIVER_CHECK(nvrtc().cuFuncSetAttribute(
+        func, CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES, sharedMemBytes));
+  }
+  return func;
+}
+
+template <size_t NUM_ARGS>
+static inline void launchKernel(
+    CUfunction func,
+    uint32_t gridX,
+    uint32_t gridY,
+    uint32_t gridZ,
+    uint32_t numWarps,
+    uint32_t sharedMemBytes,
+    std::array<void*, NUM_ARGS>& args,
+    cudaStream_t stream) {
+  AT_CUDA_DRIVER_CHECK(nvrtc().cuLaunchKernel(
+      func,
+      gridX,
+      gridY,
+      gridZ,
+      32 * numWarps,
+      1,
+      1,
+      sharedMemBytes,
+      stream,
+      args.data(),
+      nullptr));
+}
+
+/**
+   Given a list of args and their types (in a string), along with two stack
+   allocated arrays, puts each argument arg_{i} into argStorage[i], and a
+   pointer to the argument in kernelArgs[i]. We then can pass `kernelArgs`
+   directly to launchKernel. Note that some args can be less than 8 bytes, but
+   we'll still allocate 8 bytes on the stack for them.
+
+   * TODO: Need to handle NvtmDesc here.
+ */
+template <size_t NUM_ARGS>
+void parseKernelArgs(
+    PyObject* varArgs,
+    const char* argTypes,
+    std::array<uint64_t, NUM_ARGS>& argStorage,
+    std::array<void*, NUM_ARGS>& kernelArgs) {
```

Comment:
This function doesn't need to be templatized:
```
   uint64_t* argStorage,
   void* kernelArgs
```
and pass `argStorage.data()`, etc below.